demo/admin/src/pages/Page.tsx

```diff
@@ -77,4 +80,44 @@ export const Page: DocumentInterface<Pick<GQLPage, "content" | "seo">, GQLPageIn
 
         return newOutput;
     },
+    getUrl: async ({ rootColumnName, jsonPath, contentScopeUrl, apolloClient, id }) => {
```

Comment:
As this DependencyInterface is used together with other interfaces, this methods needs a more specific name: `getDependencyUrl`, or `getDependencyEditUrl`, or `getDependencyTargetUrl`, and maybe replace `get` with `fetch`.

Reply:
> As this DependencyInterface is used together with other interfaces, this methods needs a more specific name: `getDependencyUrl`, or `getDependencyEditUrl`, or `getDependencyTargetUrl`, and maybe replace `get` with `fetch`.

Really? I think, `getUrl()` isn't really specific to the dependencies. It simply provides an URL to edit a specific entity (or block). 
For example, I'd love to have a link from the image block to the image in the DAM. We could also use the method for that.

Maybe we should make it more dynamic and rename the arguments instead of renaming the method:

```
getUrl({ 
    id: string;
    apolloClient: ApolloClient<unknown>;
    contentScopeUrlPrefix: string; // instead of contentScopeUrl
    fieldName?: string; // instead of rootColumnName, now optional
    pathToBlock?: string; // instead of jsonPath, now optional
})
```

But I'm not 100% sure about the rename, maybe that makes it only more confusing.

Reply:
> and those 40 LOC need to be simplified, they should not be required in every application for every page type

That's not easy.

If you look at the getUrl() methods of different entities, they are very different:

<details>
 <summary>Page</summary>

```ts
getUrl: async ({ rootColumnName, jsonPath, contentScopeUrl, apolloClient, id }) => {
    const { data, error } = await apolloClient.query<GQLPageDependencyQuery, GQLPageDependencyQueryVariables>({
        query: gql`
            query PageDependency($id: ID!) {
                page(id: $id) {
                    id
                    content
                    seo
                    pageTreeNode {
                        id
                        category
                    }
                }
            }
        `,
        variables: {
            id,
        },
    });

    if (error || data.page.pageTreeNode === null) {
        throw new Error(`Page.getUrl: Could not find a PageTreeNode for Page with id ${id}`);
    }

    let dependencyRoute: string;
    if (rootColumnName === "content") {
        dependencyRoute = PageContentBlock.resolveDependencyRoute(
            PageContentBlock.input2State(data.page.content),
            jsonPath.substring("root.".length),
        );
    } else {
        dependencyRoute = `config/${SeoBlock.resolveDependencyRoute(SeoBlock.input2State(data.page.seo), jsonPath.substring("root.".length))}`;
    }

    return `${contentScopeUrl}/pages/pagetree/${categoryToUrlParam(data.page.pageTreeNode.category)}/${
        data.page.pageTreeNode.id
    }/edit/${dependencyRoute}`;
},
```
</details> 

<details>
 <summary>NewsDependency</summary>

```ts
getUrl: async ({ contentScopeUrl, apolloClient, id, rootColumnName, jsonPath }) => {
    const { data, error } = await apolloClient.query<GQLNewsDependencyQuery, GQLNewsDependencyQueryVariables>({
        query: gql`
            query NewsDependency($id: ID!) {
                news(id: $id) {
                    id
                    content
                }
            }
        `,
        variables: {
            id,
        },
    });

    if (error) {
        throw new Error(`News.getUrl: Could not find a News with id ${id}`);
    }

    let dependencyRoute = "";
    if (rootColumnName === "content") {
        dependencyRoute = `form/${NewsContentBlock.resolveDependencyRoute(
            NewsContentBlock.input2State(data.news.content),
            jsonPath.substring("root.".length),
        )}`;
    }

    return `${contentScopeUrl}/structured-content/news/${data.news.id}/edit/${dependencyRoute}`;
},
```
</details> 

<details>
 <summary>DamFileDependency</summary>

```ts
getUrl: async ({ contentScopeUrl, apolloClient, id }) => {
    const { data, error } = await apolloClient.query<GQLFileDependencyQuery, GQLFileDependencyQueryVariables>({
        query: gql`
            query FileDependency($id: ID!) {
                damFile(id: $id) {
                    id
                    folder {
                        id
                        parents {
                            id
                        }
                    }
                }
            }
        `,
        variables: {
            id,
        },
    });

    if (error) {
        throw new Error(`DamFile.getUrl: Could not find a DamFile with id ${id}`);
    }

    const folderIds: string[] = [];
    if (data.damFile.folder) {
        data.damFile.folder.parents.forEach((parent) => {
            folderIds.push(parent.id);
        });
        folderIds.push(data.damFile.folder.id);
    }

    return `${contentScopeUrl}/assets/${folderIds.map((id) => `${id}/folder`).join("/")}${folderIds.length > 0 ? "/" : ""}${
        data.damFile.id
    }/edit`;
},
```
</details> 

The problem is

1. The queries are completely different
2. The URL always has a custom part in the middle that must be defined manually

## Simplification ideas:

### A helper only for documents (similar to https://github.com/vivid-planet/comet/pull/1324)

Downside:
- Only works for documents
- Query must still be defined for each document
- We hardcode the URL of the PageTree in the library (but the URL is defined in the project, though it will usually be identical to the Starter)

### Generating getUrl() with the Admin Generator

Problem:

All URLs start with the contentScopeUrl, then there is a custom part, then the dependencyRoute follows (if there is any). We can never know the custom part in the Generator because you still have to manually register the route. Otherwise, we could generate the whole method.

However, we could generate a helper that contains everything except the middle part of the URL

Reply:
> The queries are completely different

- Page
- NewsDependency
- DamFileDependency: special case we can ignore

-> we possibly need two different solutions for those two use cases

> The URL always has a custom part in the middle that must be defined manually

yes, that is tricky. We should discuss that in our next meeting.